### PR TITLE
Add asm goto support to `asm!`

### PR DESCRIPTION
**Pull request by nbdd0121**
Tracking issue: #119364

This PR implements asm-goto support, using the syntax described in "future possibilities" section of [RFC2873](https://rust-lang.github.io/rfcs/2873-inline-asm.html#asm-goto).

Currently I have only implemented the `label` part, not the `fallthrough` part (i.e. fallthrough is implicit). This doesn't reduce the expressive though, since you can use label-break to get arbitrary control flow or simply set a value and rely on jump threading optimisation to get the desired control flow. I can add that later if deemed necessary.

r? @Amanieu 
cc @ojeda

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 1]
```diff
@@ -224,13 +224,15 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+            InlineAsm { template, ref operands, options, ref targets, .. } => {
                 M::eval_inline_asm(self, template, operands, options)?;
                 if options.contains(InlineAsmOptions::NORETURN) {
                     throw_ub_custom!(fluent::const_eval_noreturn_asm_returned);
                 }
                 self.go_to_block(
-                    destination
+                    targets
+                        .get(0)
+                        .copied()
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:235]: This should instead just not do `go_to_block` at all and leave it to `eval_inline_asm` to do the jump.

Please also add a doc comment [here](https://github.com/rust-lang/rust/blob/4d93590d59ca7d71ac401ce600fd0fe458a6375e/compiler/rustc_const_eval/src/interpret/machine.rs#L376) saying that the function is expected to handle the jumping to the next block.

**Review by Amanieu — COMMENTED** [review]
This needs a page in the unstable book. It should contain everything that needs to be added to the inline asm page in the reference when this is eventually stabilized.

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 2]
```diff
@@ -915,6 +916,10 @@ pub enum InlineAsmOperand<'tcx> {
     SymStatic {
         def_id: DefId,
     },
+    Label {
+        /// This represents the index into the `targets` array in `TerminatorKind::InlineAsm`.
+        target_index: usize,
```
[the remark this answers is not in the document]
Comment (nbdd0121) [reply compiler/rustc_middle/src/mir/syntax.rs:921]: It can be 0 if noreturn is used.

File: compiler/rustc_codegen_llvm/src/asm.rs [hunk 3]
```diff
@@ -457,8 +464,10 @@ pub(crate) fn inline_asm_call<'ll>(
                 can_throw,
             );
 
-            let call = if let Some((dest, catch, funclet)) = dest_catch_funclet {
-                bx.invoke(fty, None, None, v, inputs, dest, catch, funclet)
+            let call = if !labels.is_empty() {
+                bx.callbr(fty, None, None, v, inputs, dest.unwrap(), labels, None)
```
Comment (Amanieu) [inline compiler/rustc_codegen_llvm/src/asm.rs:468]: Add a `bug!` here if `catch_funclet` is not `None` for future-proofing.

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 4]
```diff
@@ -264,7 +264,7 @@ impl<'a, 'tcx> TerminatorCodegenHelper<'tcx> {
             mir::UnwindAction::Unreachable => None,
         };
 
-        if let Some(cleanup) = unwind_target {
+        if operands.iter().any(|x| matches!(x, InlineAsmOperandRef::Label { .. })) {
             let ret_llbb = if let Some(target) = destination {
```
Comment (Amanieu) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:268]: Here too.

File: src/doc/unstable-book/src/language-features/asm-goto.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    asm!(
+        "jmp {}",
+        label {
+            println!("Jumped from asm!");
+        }
+    );
+}
+```
+
+The block must have unit type.
```
Comment (Amanieu) [inline src/doc/unstable-book/src/language-features/asm-goto.md:24]: This should be updated to clarify the behavior of diverging blocks, and how this interacts with `noreturn`.

File: tests/ui/asm/x86_64/goto.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                value = true;
+            }
+        );
+        assert!(value);
+    }
+}
+
+// asm goto with outputs cause miscompilation in LLVM. UB can be triggered
+// when outputs are used inside the label block when optimisation is enabled.
+// See: https://github.com/llvm/llvm-project/issues/74483
```
Comment (Amanieu) [inline tests/ui/asm/x86_64/goto.rs:38]: This is a blocker for stabilizing this feature.

**Comment by Amanieu** [comment]
r=me

**Comment by nbdd0121** [comment]
CI is passing now

**Comment by nbdd0121** [comment]
Sorry, mistakenly committed library/stdarch..

**Pull request merged**